_core/smart/clarityRef.md

```diff
@@ -0,0 +1,193 @@
+
+### Buffer type
+
+Currently, you cannot create a `Buffer` type. They are used in returns, all of the hash functions:
+
+* `hash160`
+* `sha256`
+* `keccak256`
+
+The properties are all, themselves, buffers `time`, `header-hash`, `burnchain-header-hash`, and `vrf-seed`. 
+
+### List type
+
+Clarity supports lists of the atomic types. However, the only variable length lists in the language appear as function inputs; there is no support for list operations like append or join.
+
+### Principal type
+
+Clarity provides this primitive for checking whether or not the smart contract transaction was signed by a particular principal. Principals represent a spending entity and are roughly equivalent to a Stacks address. The principal's signature is not checked by the smart contract, but by the virtual machine. A smart contract function can use the  globally defined `tx-sender` variable to obtain the current principal.
+
+Smart contracts may also be principals (represented by the smart contract's identifier). However, there is no private key associated with the smart contract, and it cannot broadcast a signed transaction on the blockchain. A smart contract uses the special variable `contract-principal` to refer to its own principal.
+
+You can use the `is-contract?` to determine whether a given principal corresponds to a smart contract.
+
+### String type
+
+To create strings, you can use of the hash functions, `hash160`, `sha256`, and `keccak256`, for example:
+
+```cl
+(hash160 "aaron.id")
+```
+
+### Tuple type
+
+To support the use of named fields in keys and values, Clarity  allows the construction of named tuples using a function `(tuple ...)`, for example
+
+```cl
+(define imaginary-number-a (tuple #real 1 #i 2))
+(define imaginary-number-b (tuple #real 2 #i 3))
+```
+
+This allows for creating named tuples on the fly, which is useful for data maps where the keys and values are themselves named tuples. Values in a given mapping are set or fetched using:
+
+<table class="uk-table uk-table-small">
+<tr>
+  <th class="uk-width-small">Function</th>
+  <th>Description</th>
+</tr>
+<tr>
+  <td><code>(fetch-entry map-name key-tuple)</code></td>
+  <td>Fetches the value associated with a given key in the map, or returns <code>'null</code> if there is no such value.</td>
+</tr>
+<tr>
+   <td><code>(set-entry! map-name key-tuple value-tuple)</code></td>
+  <td>Sets the value of key-tuple in the data map</td>
+</tr>
+  <tr>
+   <td><code>(insert-entry! map-name key-tuple value-tuple)</code></td>
+  <td>Sets the value of key-tuple in the data map if and only if an entry does not already exist.</td>
+</tr>
+  <tr>
+   <td><code>(delete-entry! map-name key-tuple)</code></td>
+  <td>Deletes key-tuple from the data map.</td>
+</tr>
+</table>
+
+
+To access a named value of a given tuple, the `(get #name tuple)` function returns that item from the tuple.
+
+### Optional type
+
+Represents an optional value. Use the optional type for:
```

Comment:
This might be more clear:

"Represents an optional value. This is used in place of the typical usage of "null" values in other languages, and represents a type that can either be _some value_ or _none_". These types are used as the return types of data-map functions.

Reply:
🙏